gateway/src/main/java/io/zeebe/gateway/impl/job/CreateWorkflowHandler.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.gateway.impl.job;
+
+import io.zeebe.gateway.Loggers;
+import io.zeebe.gateway.cmd.BrokerErrorException;
+import io.zeebe.gateway.impl.broker.BrokerClient;
+import io.zeebe.gateway.impl.broker.BrokerResponseConsumer;
+import io.zeebe.gateway.impl.broker.RoundRobinDispatchStrategy;
+import io.zeebe.gateway.impl.broker.cluster.BrokerTopologyManager;
+import io.zeebe.gateway.impl.broker.request.BrokerCreateWorkflowInstanceRequest;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceCreationRecord;
+import io.zeebe.protocol.record.ErrorCode;
+import io.zeebe.transport.impl.sender.NoRemoteAddressFoundException;
+import java.util.function.Consumer;
+
+public class CreateWorkflowHandler {
+
+  private final BrokerClient brokerClient;
+  private final RoundRobinDispatchStrategy roundRobinDispatchStrategy;
+
+  public CreateWorkflowHandler(
+      final BrokerClient brokerClient, final BrokerTopologyManager topologyManager) {
+    this.brokerClient = brokerClient;
+    roundRobinDispatchStrategy = new RoundRobinDispatchStrategy(topologyManager);
+  }
+
+  public void createWorkflow(
+      final int partitionsCount,
+      final BrokerCreateWorkflowInstanceRequest request,
+      final BrokerResponseConsumer<WorkflowInstanceCreationRecord> responseConsumer,
+      final Consumer<Throwable> throwableConsumer) {
+    createWorkflow(
+        request,
+        partitionIdIteratorForType(partitionsCount),
+        responseConsumer,
+        throwableConsumer,
+        null);
+  }
+
+  private void createWorkflow(
+      final BrokerCreateWorkflowInstanceRequest request,
+      final PartitionIdIterator partitionIdIterator,
+      final BrokerResponseConsumer<WorkflowInstanceCreationRecord> responseConsumer,
+      final Consumer<Throwable> throwableConsumer,
+      final Throwable lastError) {
+
+    if (partitionIdIterator.hasNext()) {
+      final int partitionId = partitionIdIterator.next();
+
+      // partitions to check
+      request.setPartitionId(partitionId);
+      brokerClient.sendRequest(
+          request,
+          responseConsumer,
+          error -> {
+            if (shouldRetry(error)) {
+              Loggers.GATEWAY_LOGGER.trace(
```

Comment:
Is this what we did in the activate jobs handler? Because now we won't know the real reason we couldn't create a workflow anymore, right? I guess we at least send the last error, which is already better than nothing. We had a short discussion on slack about it, and basically it boils down to what can the user do with the error you return? So in this sense, it's quite application specific, and you need to figure out the hierarchy of errors, e.g. which carries more information such that the user can actually do something about it (e.g. retry, retry with backoff, etc.).

Anyway, this is probably fine for now, I guess - we can figure out error handling for the 0.24 implementation, unless you see any major issues here from a client PoV.